addons/calendar/models/calendar.py

```diff
@@ -0,0 +1,1038 @@
+            return
+
+        if self.allday:  # from datetime to date
+            start_datetime = self.start_datetime or self.start
+
+            if start_datetime:
+                start = fields.Datetime.from_string(start_datetime)
+                self.start_date = fields.Date.to_string(start)
+
+            stop_datetime = self.stop_datetime or self.stop
+            if stop_datetime:
+                stop = fields.Datetime.from_string(stop_datetime)
+                self.stop_date = fields.Date.to_string(stop)
+        else:  # from date to datetime
+            tz = pytz.timezone(self.env.user.tz) if self.env.user.tz else pytz.utc
+
+            if self.start_date:
+                start = fields.Datetime.from_string(self.start_date)
+                startdate = tz.localize(start).replace(hour=8).astimezone(pytz.utc)  # Add "+hh:mm" timezone, Set 8 AM in localtime, Convert to UTC
+                self.start_datetime = fields.Datetime.to_string(startdate)
+            elif self.start:
+                self.start_datetime = self.start
+
+            if self.stop_date:
+                stop = fields.Datetime.from_string(self.stop_date)
+                stop_date = tz.localize(stop).replace(hour=18).astimezone(pytz.utc)
+
+                self.stop_datetime = fields.Datetime.to_string(stop_date)
+            elif self.stop:
+                self.stop_datetime = self.stop
+
+    @api.onchange('duration', 'start_datetime')
+    def onchange_duration(self):
+        if not (self.start_datetime and self.duration):
+            return
+        start = fields.Datetime.from_string(self.start_datetime) + timedelta(hours=self.duration)
+        self.stop_date = fields.Date.to_string(start)
+        self.stop = fields.Datetime.to_string(start)
+        self.stop_datetime = fields.Datetime.to_string(start)
+        self.start_date = self.start_datetime
+        self.start = self.start_datetime
+
+    @api.onchange('start_date')
+    def onchange_start_date(self):
+        return self.onchange_dates('start', True)
+
+    @api.onchange('stop_date')
+    def onchange_stop_date(self):
+        return self.onchange_dates('stop', True)
+
+    def onchange_dates(self, fromtype, allday=False):
+
+        """Returns duration and end date based on values passed
+        """
+
+        if self.allday != allday:
+            return
+
+        if allday:
+            if fromtype == 'start' and self.start_date:
+                self.start_datetime = self.start_date
+                self.start = self.start_date
+
+            if fromtype == 'stop' and self.stop_date:
+                self.stop_datetime = self.stop_date
+                self.stop = self.stop_date
+
+    def new_invitation_token(self):
+        return uuid.uuid4().hex
+
+    def create_attendees(self):
+        res = {}
+        CalendarAttendee = new_attendees = self.env['calendar.attendee']
+        ResPartner = self.env['res.partner']
```

Comment:
But isinstance is a different function and has different behaviour to type.
Will return true for inherited classes as well.  Make sure it is correct to
do so first.

On Mon, 4 Jan 2016 8:26 PM Mohammed Shekha <notifications@github.com> wrote:

> In addons/calendar/models/calendar.py
> <https://github.com/odoo/odoo/pull/9781#discussion_r48708652>:
>
> > +        if allday:
> > +            if fromtype == 'start' and self.start_date:
> > +                self.start_datetime = self.start_date
> > +                self.start = self.start_date
> > +
> > +            if fromtype == 'stop' and self.stop_date:
> > +                self.stop_datetime = self.stop_date
> > +                self.stop = self.stop_date
> > +
> > +    def new_invitation_token(self):
> > +        return uuid.uuid4().hex
> > +
> > +    def create_attendees(self):
> > +        res = {}
> > +        CalendarAttendee = new_attendees = self.env['calendar.attendee']
> > +        ResPartner = self.env['res.partner']
>
> Here use of ResPartner is only concatenation of records, so better to keep
> name as res_partner
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/odoo/odoo/pull/9781/files#r48708652>.
>